File: sample/swift/drawLine/Sources/BuildBox.swift
import Foundation

@available(iOS 15.0, macOS 12.0, *)
class BuildBox {
    let url = URL(string: "wss://websocket.voxelamming.com")!
    let webSocketTask: URLSessionWebSocketTask
    let textureNames = ["grass", "stone", "dirt", "planks", "bricks"]
    var roomName = ""
    var isAllowedMatrix: Int = 0
    var savedMatrices: [[Double]] = []
    var translation: [Double] = [0, 0, 0, 0, 0, 0]
    var matrixTranslation: [Double] = [0, 0, 0, 0, 0, 0]
    var frameTranslations: [[Double]] = []
    var globalAnimation: [Double] = [0, 0, 0, 0, 0, 0, 1, 0]
    var animation: [Double] = [0, 0, 0, 0, 0, 0, 1, 0]
    var boxes = [[Double]]()
    var frames = [[Double]]()
    var sentence = [String]()
    var lights = [[Double]]()
    var commands = [String]()
    var size: Double = 1.0
    var shape: String = "box"
    var isMetallic: Int = 0
    var roughness: Double = 0.5
    var isAllowedFloat: Int = 0
    var buildInterval = 0.01
    var isFraming = false
    var frameId: Int = 0

    init(roomName: String) {
        self.roomName = roomName
        webSocketTask = URLSession.shared.webSocketTask(with: url)
    }

    func clearData() {
        isAllowedMatrix = 0
        savedMatrices = []
        translation = [0, 0, 0, 0, 0, 0]
        matrixTranslation = [0, 0, 0, 0, 0, 0]
        frameTranslations = []
        globalAnimation = [0, 0, 0, 0, 0, 0, 1, 0]
        animation = [0, 0, 0, 0, 0, 0, 1, 0]
        boxes = []
        frames = []
        sentence = []
        lights = []
        commands = []
        size = 1.0
        shape = "box"
        isMetallic = 0
        roughness = 0.5
        isAllowedFloat = 0
        buildInterval = 0.01
        isFraming = false
        frameId = 0
    }

    func setFrameFPS(_ fps: Int = 2) {
        commands.append("fps \(fps)")
    }

    func setFrameRepeats(_ repeats: Int = 10) {
        commands.append("repeats \(repeats)")
    }

    func frameIn() {
        isFraming = true
    }

    func frameOut() {
        isFraming = false
        frameId += 1
    }

    func pushMatrix() {
        self.isAllowedMatrix += 1
        self.savedMatrices.append(matrixTranslation)
    }

    func popMatrix() {
        self.isAllowedMatrix -= 1
        matrixTranslation = self.savedMatrices.popLast()!
    }

    func translate(_ x: Double, _ y: Double, _ z: Double, pitch: Double = 0, yaw: Double = 0, roll: Double = 0) {
        if self.isAllowedMatrix > 0 {
            // Retrieve the saved matrix
            let matrix = self.savedMatrices.last!
            let basePosition = Array(matrix[0...2])

            let baseRotationMatrix: [[Double]]
            if matrix.count == 6 {
                baseRotationMatrix = getRotationMatrix(pitch: matrix[3], yaw: matrix[4], roll: matrix[5])
            } else {
                baseRotationMatrix = [
                    Array(matrix[3...5]),
                    Array(matrix[6...8]),
                    Array(matrix[9...11])
                ]
            }

            // Calculate the position after the move
            let (addX, addY, addZ) = transformPointByRotationMatrix(point: (Double(x), Double(y), Double(z)), R: transpose3x3(matrix: baseRotationMatrix))
            let addedVectorList = addVectors(vector1: basePosition, vector2: [Double(addX), Double(addY), Double(addZ)])
            let newX = addedVectorList[0]
            let newY = addedVectorList[1]
            let newZ = addedVectorList[2]
            let roundNumList = roundNumbers(numList: [Double(newX), Double(newY), Double(newZ)])
            let roundX = roundNumList[0]
            let roundY = roundNumList[1]
            let roundZ = roundNumList[2]

            // Calculate the rotation after the move
            // inverse rotation
            let translateRotationMatrix = getRotationMatrix(pitch: -pitch, yaw: -yaw, roll: -roll)
            let rotatedMatrix = matrixMultiply(A: translateRotationMatrix, B: baseRotationMatrix)

            matrixTranslation = [roundX, roundY, roundZ] + rotatedMatrix.flatMap { $0 }
        } else {
            let roundNumList = roundNumbers(numList: [x, y, z])
            let roundX = roundNumList[0]
            let roundY = roundNumList[1]
            let roundZ = roundNumList[2]

            if isFraming {
                frameTranslations.append([roundX, roundY, roundZ, pitch, yaw, roll, Double(frameId)])
            } else {
                self.translation = [roundX, roundY, roundZ, pitch, yaw, roll]
            }
        }
    }

    func createBox(_ x: Double, _ y: Double, _  z: Double, r: Double = 1, g: Double = 1, b: Double = 1, alpha: Double = 1, texture: String = "") {
        var x = x
        var y = y
        var z = z

        if self.isAllowedMatrix > 0 {
            // Retrieve the saved matrix
            let matrix = matrixTranslation
            let basePosition = Array(matrix[0...2])

            let baseRotationMatrix: [[Double]]
            if matrix.count == 6 {
                baseRotationMatrix = getRotationMatrix(pitch: matrix[3], yaw: matrix[4], roll: matrix[5])
            } else {
                baseRotationMatrix = [
                    Array(matrix[3...5]),
                    Array(matrix[6...8]),
                    Array(matrix[9...11])
                ]
            }

            // Calculate the position after the move
            let (addX, addY, addZ) = transformPointByRotationMatrix(point: (Double(x), Double(y), Double(z)), R: transpose3x3(matrix: baseRotationMatrix))
            let addedVectorList = addVectors(vector1: basePosition, vector2: [Double(addX), Double(addY), Double(addZ)])
            x = addedVectorList[0]
            y = addedVectorList[1]
            z = addedVectorList[2]
        }

        let roundNumList = roundNumbers(numList: [x, y, z])
        let roundX = roundNumList[0]
        let roundY = roundNumList[1]
        let roundZ = roundNumList[2]
        let roundColorList = roundColors(numList: [r, g, b, alpha])
        let roundR = roundColorList[0]
        let roundG = roundColorList[1]
        let roundB = roundColorList[2]
        let roundAlpha = roundColorList[3]

        // 重ねて置くことを防止するために、同じ座標の箱があれば削除する
        removeBox(roundX, roundY, roundZ)

        let textureId: Int
        if textureNames.contains(texture) {
            textureId = textureNames.firstIndex(of: texture) ?? -1
        } else {
            textureId = -1
        }

        if isFraming {
            frames.append([roundX, roundY, roundZ, roundR, roundG, roundB, roundAlpha, Double(textureId), Double(frameId)])
        } else {
            boxes.append([x, y, z, r, g, b, alpha, Double(textureId)])
        }
    }

    func removeBox(_ x: Double, _ y: Double, _ z: Double) {
        let roundNumList = roundNumbers(numList: [x, y, z])
        let roundX = roundNumList[0]
        let roundY = roundNumList[1]
        let roundZ = roundNumList[2]

        if isFraming {
            frames.removeAll { frame in
                frame[0] == roundX && frame[1] == roundY && frame[2] == roundZ && frame[8] == Double(frameId)
            }
        } else {
            boxes.removeAll { box in
                box[0] == roundX && box[1] == roundY && box[2] == roundZ
            }
        }
    }

    func animateGlobal(_ x: Double, _ y: Double, _  z: Double, pitch: Double = 0, yaw: Double = 0, roll: Double = 0, scale: Double = 1, interval: Double = 10) {
        let roundNumList = roundNumbers(numList: [x, y, z])
        let roundX = roundNumList[0]
        let roundY = roundNumList[1]
        let roundZ = roundNumList[2]
        globalAnimation = [roundX, roundY, roundZ, pitch, yaw, roll, scale, interval]
    }

    func animate(_ x: Double, _ y: Double, _  z: Double, pitch: Double = 0, yaw: Double = 0, roll: Double = 0, scale: Double = 1, interval: Double = 10) {
        let roundNumList = roundNumbers(numList: [x, y, z])
        let roundX = roundNumList[0]
        let roundY = roundNumList[1]
        let roundZ = roundNumList[2]
        animation = [roundX, roundY, roundZ, pitch, yaw, roll, scale, interval]
    }

    func setBoxSize(_ boxSize: Double) {
        size = boxSize
    }

    func setBuildInterval(_ interval: Double) {
        buildInterval = interval
    }

    func writeSentence(_ string_sentence: String, _ x: Double, _ y: Double, _  z: Double, r: Double = 0, g: Double = 0, b: Double = 0, alpha: Double = 1) {
        let roundNumList = roundNumbers(numList: [x, y, z])
        let roundX = roundNumList[0]
        let roundY = roundNumList[1]
        let roundZ = roundNumList[2]
        let roundColorList = roundColors(numList: [r, g, b, alpha])
        let roundR = roundColorList[0]
        let roundG = roundColorList[1]
        let roundB = roundColorList[2]
        let stringX = String(roundX)
        let stringY = String(roundY)
        let stringZ = String(roundZ)
        let stringR = String(roundR)
        let stringG = String(roundG)
        let stringB = String(roundB)
        let stringAlpha = String(alpha)
        sentence = [string_sentence, stringX, stringY, stringZ, stringR, stringG, stringB, stringAlpha]
    }

    func setLight(_ x: Double, _ y: Double, _  z: Double, r: Double = 0, g: Double = 0, b: Double = 0, alpha: Double = 1, intensity: Double = 1000, interval: Double = 1, lightType: String = "point") {
        let roundNumList = roundNumbers(numList: [x, y, z])
        let roundX = roundNumList[0]
        let roundY = roundNumList[1]
        let roundZ = roundNumList[2]
        let roundColorList = roundColors(numList: [r, g, b, alpha])
        let roundR = roundColorList[0]
        let roundG = roundColorList[1]
        let roundB = roundColorList[2]
        let roundAlpha = roundColorList[3]
        var doubleLightType: Double

        if lightType == "point" {
            doubleLightType = 1
        } else if lightType == "spot" {
            doubleLightType = 2
        } else if lightType == "directional" {
            doubleLightType = 3
        } else {
            doubleLightType = 1
        }
        lights.append([roundX, roundY, roundZ, roundR, roundG, roundB, roundAlpha, intensity, interval, doubleLightType])
    }

    func setCommand(_ command: String) {
        commands.append(command)

        if command == "float" {
            isAllowedFloat = 1
        }
    }

    func drawLine(_ x1: Double, _ y1: Double, _ z1: Double, _ x2: Double, _ y2: Double, _ z2: Double, r: Double = 1, g: Double = 1, b: Double = 1, alpha: Double = 1) {
        let x1 = floor(x1)
        let y1 = floor(y1)
        let z1 = floor(z1)
        let x2 = floor(x2)
        let y2 = floor(y2)
        let z2 = floor(z2)
        let diffX = x2 - x1
        let diffY = y2 - y1
        let diffZ = z2 - z1
        let maxLength = max(abs(diffX), abs(diffY), abs(diffZ))

        if diffX == 0 && diffY == 0 && diffZ == 0 {
            return
        }

        if abs(diffX) == maxLength {
            if x2 > x1 {
                for x in Int(x1)...Int(x2) {
                    let y = y1 + (Double(x) - x1) * diffY / diffX
                    let z = z1 + (Double(x) - x1) * diffZ / diffX
                    createBox(Double(x), y, z, r: r, g: g, b: b, alpha: alpha)
                }
            } else {
                for x in stride(from: Int(x1), through: Int(x2 + 1), by: -1) {
                    let y = y1 + (Double(x) - x1) * diffY / diffX
                    let z = z1 + (Double(x) - x1) * diffZ / diffX
                    createBox(Double(x), y, z, r: r, g: g, b: b, alpha: alpha)
                }
            }
        } else if abs(diffY) == maxLength {
            if y2 > y1 {
                for y in Int(y1)...Int(y2) {
                    let x = x1 + (Double(y) - y1) * diffX / diffY
                    let z = z1 + (Double(y) - y1) * diffZ / diffY
                    createBox(x, Double(y), z, r: r, g: g, b: b, alpha: alpha)
                }
            } else {
                for y in stride(from: Int(y1), through: Int(y2 + 1), by: -1) {
                    let x = x1 + (Double(y) - y1) * diffX / diffY
                    let z = z1 + (Double(y) - y1) * diffZ / diffY
                    createBox(x, Double(y), z, r: r, g: g, b: b, alpha: alpha)
                }
            }
        } else if abs(diffZ) == maxLength {
            if z2 > z1 {
                for z in Int(z1)...Int(z2) {
                    let x = x1 + (Double(z) - z1) * diffX / diffZ
                    let y = y1 + (Double(z) - z1) * diffY / diffZ
                    createBox(x, y, Double(z), r: r, g: g, b: b, alpha: alpha)
                }
            } else {
                for z in stride(from: Int(z1), through: Int(z2 + 1), by: -1) {
                    let x = x1 + (Double(z) - z1) * diffX / diffZ
                    let y = y1 + (Double(z) - z1) * diffY / diffZ
                    createBox(x, y, Double(z), r: r, g: g, b: b, alpha: alpha)
                }
            }
        }
    }


    func changeShape(_ shape: String) {
        self.shape = shape
    }

    func changeMaterial(isMetallic: Bool = false, roughness: Double = 0.5) {
        self.isMetallic = isMetallic ? 1 : 0
        self.roughness = roughness
    }

    func sendData() async throws {
        self.webSocketTask.resume()

        let date = Date()
        let dateFormatter = ISO8601DateFormatter()
        let dateString = dateFormatter.string(from: date)
        let dataDict = [
            "translation": translation,
            "frameTranslations": frameTranslations,
            "globalAnimation": globalAnimation,
            "animation": animation,
            "boxes": boxes,
            "frames": frames,
            "sentence": sentence,
            "lights": lights,
            "commands": commands,
            "size": size,
            "shape": shape,
            "interval": buildInterval,
            "isMetallic": isMetallic,
            "roughness": roughness,
            "isAllowedFloat": isAllowedFloat,
            "date": dateString
        ] as [String : Any]

        let jsonData = try JSONSerialization.data(withJSONObject: dataDict, options: [])
        guard let jsonString = String(data: jsonData, encoding: .utf8) else {
            print("Failed to convert data to string")
            return
        }

        try await self.webSocketTask.send(.string(roomName))
        print("Joined room: \(roomName)")
        try await self.webSocketTask.send(.string(jsonString))
        print("Sent message: \(jsonString)")
    }

    private func roundNumbers(numList: [Double]) -> [Double] {
        if isAllowedFloat == 1 {
            return numList.map { round($0 * 100) / 100 }
        } else {
            return numList.map { round($0 * 10) / 10 }.map { floor($0) }
        }
    }

    private func roundColors(numList: [Double]) -> [Double] {
        return numList.map { round($0 * 100) / 100 }
    }
}

func getRotationMatrix(pitch: Double, yaw: Double, roll: Double) -> [[Double]] {
    let pitch = pitch * .pi / 180.0
    let yaw = yaw * .pi / 180.0
    let roll = roll * .pi / 180.0

    // Pitch (rotation around X-axis)
    let Rx: [[Double]] = [
        [1, 0, 0],
        [0, cos(pitch), -sin(pitch)],
        [0, sin(pitch), cos(pitch)]
    ]

    // Yaw (rotation around Y-axis)
    let Ry: [[Double]] = [
        [cos(yaw), 0, sin(yaw)],
        [0, 1, 0],
        [-sin(yaw), 0, cos(yaw)]
    ]

    // Roll (rotation around Z-axis)
    let Rz: [[Double]] = [
        [cos(roll), -sin(roll), 0],
        [sin(roll), cos(roll), 0],
        [0, 0, 1]
    ]

    // Matrix multiplication: Rx x (Rz x Ry)
    let R = matrixMultiply(A: Rx, B: matrixMultiply(A: Rz, B: Ry))
    return R
}

func matrixMultiply(A: [[Double]], B: [[Double]]) -> [[Double]] {
    var result: [[Double]] = [[0, 0, 0], [0, 0, 0], [0, 0, 0]]
    for i in 0..<3 {
        for j in 0..<3 {
            for k in 0..<3 {
                result[i][j] += A[i][k] * B[k][j]
            }
        }
    }
    return result
}

func transformPointByRotationMatrix(point: (Double, Double, Double), R: [[Double]]) -> (Double, Double, Double) {
    let (x, y, z) = point
    let x_new = R[0][0] * x + R[0][1] * y + R[0][2] * z
    let y_new = R[1][0] * x + R[1][1] * y + R[1][2] * z
    let z_new = R[2][0] * x + R[2][1] * y + R[2][2] * z
    return (x_new, y_new, z_new)
}

func addVectors(vector1: [Double], vector2: [Double]) -> [Double] {
    return zip(vector1, vector2).map(+)
}

func transpose3x3(matrix: [[Double]]) -> [[Double]] {
    return [
        [matrix[0][0], matrix[1][0], matrix[2][0]],
        [matrix[0][1], matrix[1][1], matrix[2][1]],
        [matrix[0][2], matrix[1][2], matrix[2][2]]
    ]
}

File: sample/ruby/build_box.rb
require 'json'
require 'faye/websocket'
require 'eventmachine'
require 'date'
require_relative 'matrix_util'

class BuildBox
  @@texture_names = ["grass", "stone", "dirt", "planks", "bricks"]
  @@model_names = ["Mercury", "Venus", "Earth", "Mars", "Jupiter", "Saturn", "Uranus", "Neptune", "Pluto", "Sun",
    "Moon", "ToyBiplane", "ToyCar", "Drummer", "Robot", "ToyRocket", "RocketToy1", "RocketToy2", "Skull"]

  def initialize(room_name)
    @room_name = room_name
    @is_allowed_matrix = 0
    @saved_matrices = []
    @node_transform = [0, 0, 0, 0, 0, 0]
    @matrix_transform = [0, 0, 0, 0, 0, 0]
    @frame_transforms = []
    @global_animation = [0, 0, 0, 0, 0, 0, 1, 0]
    @animation = [0, 0, 0, 0, 0, 0, 1, 0]
    @boxes = []
    @frames = []
    @sentence = []
    @lights = []
    @commands = []
    @models = []
    @model_moves = []
    @size = 1
    @shape = 'box'
    @is_metallic = 0
    @roughness = 0.5
    @is_allowed_float = 0
    @build_interval = 0.01
    @is_framing = false
    @frame_id = 0
  end

  def clear_data
    @is_allowed_matrix = 0
    @saved_matrices = []
    @node_transform = [0, 0, 0, 0, 0, 0]
    @matrix_transform = [0, 0, 0, 0, 0, 0]
    @frame_transforms = []
    @global_animation = [0, 0, 0, 0, 0, 0, 1, 0]
    @animation = [0, 0, 0, 0, 0, 0, 1, 0]
    @boxes = []
    @frames = []
    @sentence = []
    @lights = []
    @commands = []
    @models = []
    @model_moves = []
    @size = 1
    @shape = 'box'
    @is_metallic = 0
    @roughness = 0.5
    @is_allowed_float = 0
    @build_interval = 0.01
    @is_framing = false
    @frame_id = 0
  end

  def set_frame_fps(fps = 2)
    @commands << "fps #{fps}"
  end

  def set_frame_repeats(repeats = 10)
    @commands << "repeats #{repeats}"
  end

  def frame_in
    @is_framing = true
  end

  def frame_out
    @is_framing = false
    @frame_id += 1
  end

  def push_matrix
    @is_allowed_matrix += 1
    @saved_matrices.push(@matrix_transform)
  end

  def pop_matrix
    @is_allowed_matrix -= 1
    @matrix_transform = @saved_matrices.pop
  end

  def transform(x, y, z, pitch: 0, yaw: 0, roll: 0)
    if @is_allowed_matrix > 0
      # 移動用のマトリックスを計算する
      matrix = @saved_matrices.last
      puts "before matrix: #{matrix}"
      base_position = matrix[0..2]

      if matrix.length == 6
        base_rotation_matrix = get_rotation_matrix(matrix[3], matrix[4], matrix[5])
      else
        base_rotation_matrix = [matrix[3..5], matrix[6..8], matrix[9..11]]
      end

      # Compute the new position after transform
      add_x, add_y, add_z = transform_point_by_rotation_matrix([x, y, z], transpose_3x3(base_rotation_matrix))
      x, y, z = add_vectors(base_position, [add_x, add_y, add_z])
      x, y, z = round_numbers([x, y, z])

      # Compute the rotation after transform
      transform_rotation_matrix = get_rotation_matrix(-pitch, -yaw, -roll)
      rotate_matrix = matrix_multiply(transform_rotation_matrix, base_rotation_matrix)
      @matrix_transform = [x, y, z, *rotate_matrix[0], *rotate_matrix[1], *rotate_matrix[2]]
    else
      x, y, z = round_numbers([x, y, z])

      if @is_framing
        @frame_transforms.append([x, y, z, pitch, yaw, roll, @frame_id])
      else
        @node_transform = [x, y, z, pitch, yaw, roll]
      end
    end
  end

  def create_box(x, y, z, r: 1, g: 1, b: 1, alpha: 1, texture: '')
    if @is_allowed_matrix > 0
      # 移動用のマトリックスにより位置を計算する
      matrix = @matrix_transform
      base_position = matrix[0..2]

      if matrix.length == 6
        base_rotation_matrix = get_rotation_matrix(matrix[3], matrix[4], matrix[5])
      else
        base_rotation_matrix = [matrix[3..5], matrix[6..8], matrix[9..11]]
      end

      # Compute the new position after transform
      add_x, add_y, add_z = transform_point_by_rotation_matrix([x, y, z], transpose_3x3(base_rotation_matrix))
      x, y, z = add_vectors(base_position, [add_x, add_y, add_z])
    end

    x, y, z = round_numbers([x, y, z])
    r, g, b, alpha = round_two_decimals([r, g, b, alpha])

    # 重ねておくことを防止
    remove_box(x, y, z)
    if !@@texture_names.include?(texture)
      texture_id = -1
    else
      texture_id = @@texture_names.index(texture)
    end

    if @is_framing
      @frames.append([x, y, z, r, g, b, alpha, texture_id, @frame_id])
    else
      @boxes.append([x, y, z, r, g, b, alpha, texture_id])
    end
  end

  def remove_box(x, y, z)
    x, y, z = round_numbers([x, y, z])

    if @is_framing
      @frames.reject! { |frame| frame[0] == x && frame[1] == y && frame[2] == z && frame[8] == @frame_id }
    else
      @boxes.reject! { |box| box[0] == x && box[1] == y && box[2] == z }
    end
  end

  def animate_global(x, y, z, pitch: 0, yaw: 0, roll: 0, scale: 1, interval: 10)
    x, y, z = round_numbers([x, y, z])
    @global_animation = [x, y, z, pitch, yaw, roll, scale, interval]
  end

  def animate(x, y, z, pitch: 0, yaw: 0, roll: 0, scale: 1, interval: 10)
    x, y, z = round_numbers([x, y, z])
    @animation = [x, y, z, pitch, yaw, roll, scale, interval]
  end

  def set_box_size(box_size)
    @size = box_size
  end

  def set_build_interval(interval)
    @build_interval = interval
  end

  def write_sentence(sentence, x, y, z, r: 1, g: 1, b: 1, alpha: 1)
    x, y, z = round_numbers([x, y, z]).map(&:to_s)
    r, g, b, alpha = round_two_decimals([r, g, b, alpha])
    r, g, b, alpha =  [r, g, b, alpha].map(&:floor).map(&:to_s)
    @sentence = [sentence, x, y, z, r, g, b, alpha]
  end

  def set_light(x, y, z, r: 1, g: 1, b: 1, alpha: 1, intensity: 1000, interval: 1, light_type: 'point')
    x, y, z = round_numbers([x, y, z])
    r, g, b, alpha = round_two_decimals([r, g, b, alpha])

    if light_type == 'point'
      light_type = 1
    elsif light_type == 'spot'
      light_type = 2
    elsif light_type == 'directional'
      light_type = 3
    else
      light_type = 1
    end
    @lights << [x, y, z, r, g, b, alpha, intensity, interval, light_type]
  end

  def set_command(command)
    @commands << command

    if command == 'float'
      @is_allowed_float = 1
    end
  end

  def draw_line(x1, y1, z1, x2, y2, z2, r: 1, g: 1, b: 1, alpha: 1)
    x1, y1, z1, x2, y2, z2 = [x1, y1, z1, x2, y2, z2].map(&:floor)
    diff_x = x2 - x1
    diff_y = y2 - y1
    diff_z = z2 - z1
    max_length = [diff_x.abs, diff_y.abs, diff_z.abs].max

    return false if diff_x == 0 && diff_y == 0 && diff_z == 0

    if diff_x.abs == max_length
      if x2 > x1
        (x1..x2).each do |x|
          y = y1 + (x - x1) * diff_y.to_f / diff_x
          z = z1 + (x - x1) * diff_z.to_f / diff_x
          create_box(x, y, z, r: r, g: g, b: b, alpha: alpha)
        end
      else
        x1.downto(x2 + 1) do |x|
          y = y1 + (x - x1) * diff_y.to_f / diff_x
          z = z1 + (x - x1) * diff_z.to_f / diff_x
          create_box(x, y, z, r: r, g: g, b: b, alpha: alpha)
        end
      end
    elsif diff_y.abs == max_length
      if y2 > y1
        (y1..y2).each do |y|
          x = x1 + (y - y1) * diff_x.to_f / diff_y
          z = z1 + (y - y1) * diff_z.to_f / diff_y
          create_box(x, y, z, r: r, g: g, b: b, alpha: alpha)
        end
      else
        y1.downto(y2 + 1) do |y|
          x = x1 + (y - y1) * diff_x.to_f / diff_y
          z = z1 + (y - y1) * diff_z.to_f / diff_y
          create_box(x, y, z, r: r, g: g, b: b, alpha: alpha)
        end
      end
    elsif diff_z.abs == max_length
      if z2 > z1
        (z1..z2).each do |z|
          x = x1 + (z - z1) * diff_x.to_f / diff_z
          y = y1 + (z - z1) * diff_y.to_f / diff_z
          create_box(x, y, z, r: r, g: g, b: b, alpha: alpha)
        end
      else
        z1.downto(z2 + 1) do |z|
          x = x1 + (z - z1) * diff_x.to_f / diff_z
          y = y1 + (z - z1) * diff_y.to_f / diff_z
          create_box(x, y, z, r: r, g: g, b: b, alpha: alpha)
        end
      end
    end
  end

  def change_shape(shape)
    @shape = shape
  end

  def change_material(is_metallic: false, roughness: 0.5)
    @is_metallic = is_metallic ? 1 : 0
    @roughness = roughness
  end

  def create_model(model_name, x: 0, y: 0, z: 0, pitch: 0, yaw: 0, roll: 0, scale: 1, entity_name: '')
    if @@model_names.include?(model_name)
      x, y, z, pitch, yaw, roll, scale = round_two_decimals([x, y, z, pitch, yaw, roll, scale])
      x, y, z, pitch, yaw, roll, scale = [x, y, z, pitch, yaw, roll, scale].map(&:to_s)

      @models << [model_name, x, y, z, pitch, yaw, roll, scale, entity_name]
    else
      puts "No model name: #{model_name}"
    end
  end

  def move_model(entity_name, x: 0, y: 0, z: 0, pitch: 0, yaw: 0, roll: 0, scale: 1)
    x, y, z, pitch, yaw, roll, scale = round_two_decimals([x, y, z, pitch, yaw, roll, scale])
    x, y, z, pitch, yaw, roll, scale = [x, y, z, pitch, yaw, roll, scale].map(&:to_s)

    @model_moves << [entity_name, x, y, z, pitch, yaw, roll, scale]
  end


  def send_data(name: '')
    puts 'send_data'
    now = DateTime.now
    data_to_send = {
      "nodeTransform": @node_transform,
      "frameTransforms": @frame_transforms,
      "globalAnimation": @global_animation,
      "animation": @animation,
      "boxes": @boxes,
      "frames": @frames,
      "sentence": @sentence,
      "lights": @lights,
      "commands": @commands,
      "models": @models,
      "modelMoves": @model_moves,
      "size": @size,
      "shape": @shape,
      "interval": @build_interval,
      "isMetallic": @is_metallic,
      "roughness": @roughness,
      "isAllowedFloat": @is_allowed_float,
      "name": name,
      "date": now.to_s
    }.to_json

    EM.run do
      ws = Faye::WebSocket::Client.new('wss://websocket.voxelamming.com')

      ws.on :open do |_event|
        p [:open]
        puts 'WebSocket connection open'
        ws.send(@room_name)
        puts "Joined room: #{@room_name}"
        ws.send(data_to_send)
        puts data_to_send
        puts 'Sent data to server'

        EM.add_timer(1) do
          ws.close
          EM.stop
        end
      end

      ws.on :error do |event|
        puts "WebSocket error: #{event.message}"
        EM.stop
      end

      ws.on :close do |_event|
        puts 'WebSocket connection closed'
        EM.stop
      end
    end
  end

  def round_numbers(num_list)
    if @is_allowed_float == 1
      round_two_decimals(num_list)
    else
      num_list.map { |val| val.round(1).floor }
    end
  end

  def round_two_decimals(num_list)
    num_list.map { |val| val.round(2) }
  end
end
